src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs

```diff
@@ -289,70 +287,78 @@ static void ThrowNodeIsNotSame(TrieNode node, TrieNode cachedNodeCopy) =>
                 throw new InvalidOperationException($"The hash of replacement node {cachedNodeCopy} is not the same as the original {node}.");
         }
 
-        public ICommitter BeginCommit(TrieType trieType, long blockNumber, Hash256? address, TrieNode? root, WriteFlags writeFlags)
+        public ICommitter BeginCommit(Hash256? address, TrieNode? root, WriteFlags writeFlags)
         {
-            ArgumentOutOfRangeException.ThrowIfNegative(blockNumber);
-            EnsureCommitSetExistsForBlock(blockNumber);
-
-            int concurrency = _pruningStrategy.PruningEnabled
-                ? Environment.ProcessorCount
-                : 0; // The write batch when pruning is not enabled is not concurrent safe
+            if (_pruningStrategy.PruningEnabled)
+            {
+                if (_currentBlockCommitter is null) throw new InvalidOperationException($"With pruning triestore, {nameof(BeginBlockCommit)} must be called.");
+            }
 
-            return new TrieStoreCommitter(this, trieType, blockNumber, address, root, writeFlags, concurrency);
+            if (_currentBlockCommitter != null)
+            {
+                // Note, archive node still use this path. This is because it need the block commit set to handle
+                // reorg announcement.
+                return _currentBlockCommitter.GetTrieCommitter(address, root, writeFlags);
+            }
+            else
+            {
+                // This happens when there are no block involved, such as during snap sync or just calculating
+                // patricia root.
+                return new NonPruningTrieStoreCommitter(this, 0, address, _nodeStorage.StartWriteBatch(), writeFlags);
+            }
```

Comment:
```suggestion
            return _currentBlockCommitter is not null 
                // Note, archive node still use this path. This is because it needs the block commit set to handle reorg announcement.
                ? _currentBlockCommitter.GetTrieCommitter(address, root, writeFlags)  
                // This happens when there are no block involved, such as during snap sync or just calculating patricia root.
                : new NonPruningTrieStoreCommitter(this, 0, address, _nodeStorage.StartWriteBatch(), writeFlags);
```